src/BeadingOrderOptimizer.cpp

```diff
@@ -0,0 +1,351 @@
+            if (polyline.junctions.empty()) 
+            {
+                continue; // Shouldn't happen 
+            }
+            polyline_end_points.emplace(polyline.junctions.front().p, ExtrusionLineEndRef(polyline.inset_idx, polyline_it, true));
+            polyline_end_points.emplace(polyline.junctions.back().p, ExtrusionLineEndRef(polyline.inset_idx, polyline_it, false));
+        }
+    }
+}
+
+
+void BeadingOrderOptimizer::fuzzyConnect(std::vector<std::list<ExtrusionLine>>& polygons_per_index, coord_t snap_dist, bool reduce_overlapping_segments, bool connect_odd_lines_to_polygons)
+{
+    struct Locator
+    {
+        Point operator()(const ExtrusionLineEndRef& it)
+        {
+            return it.p();
+        }
+    };
+    SparsePointGridInclusive<ExtrusionLineEndRef> polyline_grid(snap_dist); // Inclusive because iterators might get invalidated
+    for (std::list<ExtrusionLine>& polys : polylines_per_index)
+    {
+        for (auto poly_it = polys.begin(); poly_it != polys.end(); ++poly_it)
+        {
+            for (bool front : { true, false })
+            {
+                ExtrusionLineEndRef ref(poly_it->inset_idx, poly_it, front);
+                polyline_grid.insert(ref.p(), ref);
+            }
+        }
+    }
+    
+    struct PointLocator
+    {
+        Point operator()(const Point& p) 
+        { 
+            return p; 
+        }
+    };
+    SparsePointGrid<Point, PointLocator> polygon_grid(snap_dist); // Inclusive because iterators might get invalidated
+    for (auto polygons : polygons_per_index)
+    {
+        for (auto polygon : polygons)
+        {
+            for (ExtrusionJunction& junction : polygon.junctions)
+            {
+                polygon_grid.insert(junction.p);
+            }
+        }
+    }
+
+    // Order of end_points_to_check and nearby_end_points determines which ends are connected together
+    // TODO: decide on the best way to connect polylines at 3-way intersections
+    std::list<ExtrusionLineEndRef> end_points_to_check;
+
+    for (std::list<ExtrusionLine>& polys : polylines_per_index)
+    {
+        for (bool odd_lines : { connect_odd_lines_to_polygons, !connect_odd_lines_to_polygons })
+        { // Try to combine odd lines into polygons first, if connect_odd_lines_to_polygons
+            for (auto poly_it = polys.begin(); poly_it != polys.end(); ++poly_it)
+            {
+                if (poly_it->is_odd != odd_lines)
+                {
+                    continue;
+                }
+                
+                assert(poly_it->junctions.size() > 1);
+                for (bool front : { true, false })
+                {
+                    end_points_to_check.emplace_back(poly_it->inset_idx, poly_it, front);
+                }
+            }
+        }
+    }
+    
+    for (auto it = end_points_to_check.begin(); it != end_points_to_check.end(); ++it)
```

Comment:
Just use a for-each loop here.
```suggestion
    for (ExtrusionLineEndRef& end_point : end_points_to_check)
```